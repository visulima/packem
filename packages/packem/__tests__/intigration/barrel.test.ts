import { rm } from "node:fs/promises";
import { join } from "node:path";

import { readFileSync, writeFile, writeJson } from "@visulima/fs";
import { temporaryDirectory } from "tempy";
import { afterEach, beforeEach, describe, expect, it } from "vitest";

import { createPackageJson, createPackemConfig, execPackem } from "../helpers";

describe("barrel", () => {
    let temporaryDirectoryPath: string;

    beforeEach(async () => {
        temporaryDirectoryPath = temporaryDirectory();
    });

    afterEach(async () => {
        await rm(temporaryDirectoryPath, { recursive: true });
    });

    it("should build a project with barrel re-exports and run output", async () => {
        expect.assertions(4);

        const source = (p: string) => join(temporaryDirectoryPath, "src", p);

        await writeJson(join(temporaryDirectoryPath, "src", "package.json"), { sideEffects: false }, { overwrite: true });

        // named-barrel
        await writeFile(source("named-barrel/a.js"), "export const a = 'a';\n");
        await writeFile(source("named-barrel/b.js"), "export const b = 'b';\n");
        await writeFile(
            source("named-barrel/index.js"),
            "export { a as b } from './a';\nexport { b as c } from './b';\n",
        );

        // mixed-barrel
        await writeFile(source("mixed-barrel/a.js"), "export default 'a';\n");
        await writeFile(source("mixed-barrel/b.js"), "export const b = 'b';\n");
        await writeFile(source("mixed-barrel/c.js"), "export const value = 'c';\n");
        await writeFile(
            source("mixed-barrel/index.js"),
            "export { default as a } from './a';\nexport * from './b';\nexport { value as b } from './c';\nexport const d = 'd';\n",
        );

        // star-barrel
        await writeFile(source("star-barrel/a.js"), "export const a = 'a';\n");
        await writeFile(source("star-barrel/b.js"), "export const b = 'b';\n");
        await writeFile(source("star-barrel/c.js"), "export const c = 'c';\n");
        await writeFile(
            source("star-barrel/index.js"),
            "export { c } from './c';\nexport * from './a';\nexport * from './b';\nexport const d = 'd';\n",
        );

        // nested-barrel
        await writeFile(source("nested-barrel/b.js"), "export const b = 'b';\n");
        await writeFile(source("nested-barrel/c.js"), "export const c = 'c';\n");
        await writeFile(
            source("nested-barrel/a.js"),
            "import { b as a } from './b';\nexport { a };\nexport { c } from './c';\n",
        );
        await writeFile(
            source("nested-barrel/index.js"),
            "export { a } from './a';\nexport { c } from './c';\n",
        );

        // Entry that imports from barrels and exports a run function for assertions
        await writeFile(
            source("index.js"),
            [
                "import { b as a } from './named-barrel/index.js';",
                "import { b as c, d } from './mixed-barrel/index.js';",
                "import { b } from './star-barrel/index.js';",
                "import * as nested from './nested-barrel/index.js';",
                "export const run = () => ({ a, b, c, d, nestedA: nested.a });",
                "export default run;",
                "",
            ].join("\n"),
        );

        await createPackageJson(temporaryDirectoryPath, {
            exports: {
                ".": {
                    import: "./dist/index.mjs",
                    require: "./dist/index.cjs",
                },
            },
            sideEffects: false,
        });

        await createPackemConfig(temporaryDirectoryPath, {
            transformer: "esbuild",
        });

        const binProcess = await execPackem("build", [], {
            cwd: temporaryDirectoryPath,
            reject: false,
        });

        expect(binProcess.stderr).toBe("");
        expect(binProcess.exitCode).toBe(0);

        const mjsContent = readFileSync(
            `${temporaryDirectoryPath}/dist/index.mjs`,
        );

        expect(mjsContent).toMatchSnapshot("ESM output");

        const cjsContent = readFileSync(
            `${temporaryDirectoryPath}/dist/index.cjs`,
        );

        expect(cjsContent).toMatchSnapshot("CommonJS output");
    });

    it("should rewrite barrel imports and run output", async () => {
        expect.assertions(4);

        const source = (p: string) => join(temporaryDirectoryPath, "src", p);

        // optional: mark src as side-effect-free to align with common barrel setups
        await writeJson(join(temporaryDirectoryPath, "src", "package.json"), { sideEffects: false }, { overwrite: true });

        // leaf modules
        await writeFile(source("components/foo.js"), "export default 'foo';\n");
        await writeFile(source("components/bar.js"), "export const bar = 'bar';\n");
        await writeFile(source("components/baz.js"), "export const baz = 'baz';\n");

        // barrel file (re-exports)
        await writeFile(
            source("components/index.js"),
            [
                "export { default as Foo } from './foo';",
                "export { bar } from './bar';",
                "export * from './baz';",
                "",
            ].join("\n"),
        );

        // entry
        await writeFile(
            source("index.js"),
            [
                "import { Foo, bar, baz } from './components/index.js';",
                "export const run = () => ({ Foo, bar, baz });",
                "export default run;",
                "",
            ].join("\n"),
        );

        await createPackageJson(temporaryDirectoryPath, {
            exports: {
                ".": {
                    import: "./dist/index.mjs",
                    require: "./dist/index.cjs",
                },
            },
            sideEffects: false,
        });

        await createPackemConfig(temporaryDirectoryPath, {
            transformer: "esbuild",
        });

        const binProcess = await execPackem("build", [], {
            cwd: temporaryDirectoryPath,
            reject: false,
        });

        expect(binProcess.stderr).toBe("");
        expect(binProcess.exitCode).toBe(0);

        const mjsContent = readFileSync(
            `${temporaryDirectoryPath}/dist/index.mjs`,
        );

        expect(mjsContent).toMatchSnapshot("ESM output");

        const cjsContent = readFileSync(
            `${temporaryDirectoryPath}/dist/index.cjs`,
        );

        expect(cjsContent).toMatchSnapshot("CommonJS output");
    });
});
