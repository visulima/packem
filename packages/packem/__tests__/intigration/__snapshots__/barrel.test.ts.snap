// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`barrel > should build a project with barrel re-exports and run output > CommonJS output 1`] = `
"'use strict';

Object.defineProperties(exports, { __esModule: { value: true }, [Symbol.toStringTag]: { value: 'Module' } });

const a = "a";

const value = "c";

const d = "d";

const b$1 = "b";

const b = "b";

var __defProp = Object.defineProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
const run = /* @__PURE__ */ __name(() => ({ a, b: b$1, c: value, d, nestedA: b }), "run");

exports.default = run;
exports.run = run;
"
`;

exports[`barrel > should build a project with barrel re-exports and run output > ESM output 1`] = `
"const a = "a";

const value = "c";

const d = "d";

const b$1 = "b";

const b = "b";

var __defProp = Object.defineProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
const run = /* @__PURE__ */ __name(() => ({ a, b: b$1, c: value, d, nestedA: b }), "run");

export { run as default, run };
"
`;

exports[`barrel > should rewrite barrel imports and run output > CommonJS output 1`] = `
"'use strict';

Object.defineProperties(exports, { __esModule: { value: true }, [Symbol.toStringTag]: { value: 'Module' } });

const Foo = "foo";

const bar = "bar";

const baz = "baz";

var __defProp = Object.defineProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
const run = /* @__PURE__ */ __name(() => ({ Foo, bar, baz }), "run");

exports.default = run;
exports.run = run;
"
`;

exports[`barrel > should rewrite barrel imports and run output > ESM output 1`] = `
"const Foo = "foo";

const bar = "bar";

const baz = "baz";

var __defProp = Object.defineProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
const run = /* @__PURE__ */ __name(() => ({ Foo, bar, baz }), "run");

export { run as default, run };
"
`;
